components/authz-service/storage/postgres/migration/sql/63_make_db_id_lookup_functions_strict.up.sql

```diff
@@ -0,0 +1,109 @@
+BEGIN;
+
+-- lookup by name, the UUID column can go away
+-- Note: no need to raise a custom exception here, either, as we're only
+-- calling this function internally after having upsert'ed the members.
+CREATE OR REPLACE FUNCTION member_db_id (
+    _name iam_members.name % TYPE,
+    OUT _db_id iam_members.db_id % TYPE
+)
+    RETURNS iam_members.db_id % TYPE
+    AS $$
+BEGIN
+    SELECT
+        db_id INTO STRICT _db_id
+    FROM
+        iam_members
+    WHERE
+        name = _name;
+END;
+$$
+LANGUAGE plpgsql;
+
+-- Note: we don't raise any customized exceptions here, since this is
+-- never called using user-provided inputs. It's only called internally
+-- when linking the statements to their policies; and when deleting a
+-- policy's members.
+CREATE OR REPLACE FUNCTION policy_db_id (
+    _id iam_policies.id % TYPE,
+    OUT _db_id iam_policies.db_id % TYPE
+)
+    RETURNS iam_policies.db_id % TYPE
+    AS $$
+BEGIN
+    SELECT
+        db_id INTO STRICT _db_id
+    FROM
+        iam_policies
+    WHERE
+        id = _id;
+END;
+$$
+LANGUAGE plpgsql;
+
+CREATE OR REPLACE FUNCTION role_db_id (
+    _id iam_roles.id % TYPE,
+    OUT _db_id iam_roles.db_id % TYPE
+)
+    RETURNS iam_roles.db_id % TYPE
+    AS $$
+BEGIN
+    SELECT
+        db_id INTO _db_id
```

Comment:
Is the reason not to use strict here so that you can customize the error message?

Reply:
Not quite -- you can customize the error message with `INTO STRICT` as well:

```
BEGIN
    SELECT * INTO STRICT myrec FROM emp WHERE empname = myname;
    EXCEPTION
        WHEN NO_DATA_FOUND THEN
            RAISE EXCEPTION 'employee % not found', myname;
        WHEN TOO_MANY_ROWS THEN
            RAISE EXCEPTION 'employee % not unique', myname;
END;
```
[See here](https://www.postgresql.org/docs/9.6/plpgsql-statements.html)

Both would have the same effect (except that we don't need to worry about the _too many rows_ case, since we've got indices on those fields).

**However**, [the docs](https://www.postgresql.org/docs/9.6/plpgsql-control-structures.html) also say that having `EXCEPTION` blocks makes this more expensive in processing:

> Tip: A block containing an EXCEPTION clause is significantly more expensive to enter and exit than a block without one. Therefore, don't use EXCEPTION without need.

Asking in the PG slack's #beginners channel only yielded that `IF NOT FOUND THEN` was more common.